src/miniscript/evaluate.rs

```diff
@@ -0,0 +1,561 @@
+        witness.push(vec![]);
+        return Ok(ret)
+    }
+
+    for witness_sig in sigs {
+        for pk in keys {
+            if let Ok(()) = verify_sersig(&pk.to_public_key(), &witness_sig, msg, secp) {
+                ret.pks.push(pk.clone().to_public_key());
+                break;
+            }
+        }
+    }
+    if ret.pks.len() == k {
+        witness.push(vec![1]);
+        Ok(ret)
+    } else {
+        Err(Error::CouldnotEvaluate)
+    }
+}
+
+/// Equivalent of `n CHECKSEQUENCEVERIFY`. There is no Unsat condition for CSV, hence result is not
+/// sent using EvalResult
+fn evaluate_csv(n: u32, age: u32) -> Result<Primitives, Error>
+{
+    if age >= n {
+        Ok(Primitives{pks: vec![], hashlocks: vec![], timelocks: vec![n]})
+    } else {
+        Err(Error::LocktimeNotMet(n))
+    }
+}
+
+/// Script: DUP IF n CHECKSEQUENCEVERIFY DROP ENDIF. This does not actually check the timelock on
+/// input 0. Upon input 1, this function does CSV and DROP.
+fn evaluate_time(n: u32, age: u32, witness: &mut Vec<Vec<u8>>) -> Result<Primitives, Error>
+{
+    let top = witness.pop().expect("Witness Stack must contain one element");
+    if top == vec![]{
+        witness.push(vec![]);
+        Ok(Primitives{pks: vec![], hashlocks: vec![], timelocks: vec![]})
+    }
+    else if top == vec![1]{
+        let res = evaluate_csv(n, age)?;
+        witness.push(vec![1]);
+        Ok(res)
+    }
+    else {
+        Err(Error::CouldnotEvaluate)
+    }
+}
+
+/// helper function for calculating equivalent of HASH256 hash OP_EQUAL
+/// len(preimage) = 32 and SHA256(preimage) = hash
+fn evaluate_hash(hash: sha256::Hash, preimage: &Vec<u8>, witness: &mut Vec<Vec<u8>>)
+                   -> Result<Primitives, Error >
+{
+    if preimage.len() != 32{
+        return Err(Error::VerifyHashFail(hash, preimage.clone()))
+    }
+    if sha256::Hash::hash(preimage) == hash{
+        witness.push(vec![1]);
+        Ok(Primitives{pks: vec![], hashlocks: vec![preimage.clone()], timelocks: vec![]})
+    }
+    else{
+        Err(Error::VerifyHashFail(hash, preimage.clone()))
+    }
+}
+
+fn evaluate_hash_t(hash: sha256::Hash, witness:&mut Vec<Vec<u8>>)
+    -> Result<Primitives, Error >
+{
+    let preimage = witness.pop().expect("Preimage missing");
+    evaluate_hash(hash, &preimage, witness)
+}
+
+fn evaluate_hash_w(hash: sha256::Hash, witness: &mut Vec<Vec<u8>>)
+    -> Result<Primitives, Error >
+{
+    let witnesslen = witness.len();
+    witness.swap(witnesslen-1, witnesslen -2);//swap the last elements
+    let preimage = witness.pop().expect("Preimage missing");
+    if preimage == vec![]{
+        witness.push(vec![]);
+        return Ok(Primitives{pks: vec![], hashlocks: vec![], timelocks: vec![]})
+    }
+    evaluate_hash(hash, &preimage, witness)
+}
+
+/// A and B. Evaluates both A and B and then computes BOOLAND
+fn interpret_and_bool<P>(
+    left: &AstElem<P>,
+    right: &AstElem<P>,
+    witness: &mut Vec<Vec<u8>>,
```

Comment:
I think that the `witness`, rather than being a `Vec<Vec<u8>>`, should be renamed to `stack`, and be a vector of objects of the type
```
enum StackElement {
    /// Result of a satisfied E/T/W/F
    Satisfied,
    /// Result of a dissatisfied E/W
    Dissatisfied,
    /// Input from the witness stack
    WitnessElement(Vec<u8>),
}
```
rather than using `vec![]` or `vec[1]` the way you would with Script. This way you also eliminate the complexity of using `build_scriptint`.

Reply:
Given a witness stack, you'd go through a preprocessing stage where you mapped every empty vector to `Satisfied`, every `vec![1]` to `Dissatisfied`, and everything else to `Witness`.

This has the slightly weird effect that missing hash preimages or signatures, which are required to be `vec![]` by Miniscript, will be interpreted as "Dissatisfied", but I think that makes sense.